### ISSUE TITLE
[MNG-6386] ${project.baseUri} is not a valid URI (according to RFC 3986)

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#7482]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** opened **[MNG-6386](https://issues.apache.org/jira/browse/MNG-6386?redirect=false)** and commented

`File#toURI` produces an invalid URI on Windows:

```
file:/C:/path/to/basedir
```

A valid URI has to be:

```
file:///C:/path/to/basedir
```

Same issue for Unix-like OSes:

```
file:/path/to/basedir
```

A valid URI has to be:

```
file:///path/to/basedir
```

Using `Path#toUri` we can easily solve that problem because it creates compliant URIs. The failure occurs when interacting with local Git and Subversion repos:

```
PS D:\Entwicklung\Projekte\maven-scm> svn ls file:/D:/Entwicklung/svn-repos/scm-svn-test-at-sign
svn: E020024: Error resolving case of 'file:\D:\Entwicklung\svn-repos\scm-svn-test-at-sign'

while the proper (new) havior will produce a valid URI:
PS D:\Entwicklung\Projekte\maven-scm> svn ls file:///D:/Entwicklung/svn-repos/scm-svn-test-at-sign
branches/
tags/
trunk/ 
```

```
PS D:\Entwicklung\Projekte> git clone file:///D:/Entwicklung/git-repos/toll tlll2
Cloning into 'tlll2'...
warning: You appear to have cloned an empty repository.
PS D:\Entwicklung\Projekte> git clone file:/D:/Entwicklung/git-repos/toll tlll2
Cloning into 'tlll2'...
ssh: Could not resolve hostname file: Name or service not known
fatal: Could not read from remote repository.
```

or Subversion repo at: `D:\Entwicklung\svn-repos\это по-русский`:

```
PS D:\Entwicklung\Projekte> svn ls file:/D:/Entwicklung/svn-repos/это%20по-русский/
svn: E020024: Error resolving case of 'file:\D:\Entwicklung\svn-repos\???%20??-???????\'
```

proper URI gives:

```
PS D:\Entwicklung\Projekte> svn ls file:///D:/Entwicklung/svn-repos/%D1%8D%D1%82%D0%BE%20%D0%BF%D0%BE-%D1%80%D1%83%D1%81%D1%81%D0%BA%D0%B8%D0%B9/
branches/
tags/
trunk/
```

---

**Affects:** 3.5.3

**Issue Links:**
- [MNG-3760](https://issues.apache.org/jira/browse/MNG-3760) Support property ${baseurl} to get RFC-compliant URL of project base directory
 (_**"is caused by"**_)
- [SCM-877](https://issues.apache.org/jira/browse/SCM-877) AbstractSvnScmProvider#parseScmUrl() accepts file URLs with backslashes while SvnTagBranchUtils cannot handle them
 (_**"is depended upon by"**_)

**Remote Links:**
- [SO reference
](https://stackoverflow.com/a/17870390/696632)

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Pushed branches to maven.git and maven-integration-testing.git runs flawlessly on

```
Apache Maven 3.5.4-SNAPSHOT (fef1462102d8a3c42f31b6e43dae988767c814ff; 2018-03-30T21:54:36+02:00)
Maven home: D:\apache-maven-3.5.4-SNAPSHOT
Java version: 1.8.0_162, vendor: Oracle Corporation
Java home: C:\Program Files\Java\jdk1.8.0_162\jre
Default locale: de_DE, platform encoding: Cp1252
OS name: "windows 10", version: "10.0", arch: "amd64", family: "windows"
```

and

```
Apache Maven 3.5.4-SNAPSHOT (43b34598629f086523a333dc18665389643832a5; 2018-03-30T22:23:14+02:00)
Maven home: /home/mosipov/apache-maven-3.5.4-SNAPSHOT
Java version: 1.8.0_162, vendor: Oracle Corporation
Java home: /usr/local/openjdk8/jre
Default locale: de_DE, platform encoding: UTF-8
OS name: "freebsd", version: "11.1-release-p8", arch: "i386", family: "unix"
```

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [8e0efaa107c5dab09a9dc0a0405a3bfbebecda3b](https://gitbox.apache.org/repos/asf?p=maven.git;a=commit;h=8e0efaa107c5dab09a9dc0a0405a3bfbebecda3b)
and [63b8c5428e0e15df4c95a488b21b71aeace336ec](https://gitbox.apache.org/repos/asf?p=maven-integration-testing.git;a=commit;h=63b8c5428e0e15df4c95a488b21b71aeace336ec).
